### PR TITLE
add possibility to set a period to update link output ports

### PR DESCRIPTION
**Pull request by dbcesar**

### CODE REVIEW

**Comment by dbcesar** [comment]
@doudou I would like to try it out with the dvl_velocities, however its properties are not set directly from yml file, could you help me on that?

File: tasks/ModelTask.cpp [hunk 1]
```diff
@@ -237,9 +251,8 @@ void ModelTask::updateLinks(base::Time const& time)
         rbs.orientation = base::Quaterniond(
             source2target.rot.w,source2target.rot.x,source2target.rot.y,source2target.rot.z );
         rbs.time = time;
-
-        rbs.time = base::Time::now();
```
Comment (dbcesar) [inline tasks/ModelTask.cpp:241]: I didn't see the purpose of have this, so I took it out.

**Comment by dbcesar** [comment]
It seems that the values informed in ./config/orogen/gazebo/rock_gazebo::ModelTask.yml are not being set. Are we using them?

File: tasks/ModelTask.cpp [hunk 2]
```diff
@@ -89,13 +89,16 @@ void ModelTask::setupLinks()
...
         exported_link.port_name = it->port_name;
+        exported_link.port_period = it->port_period;
 
         if (exported_link.source_link != _world_frame.get() && !exported_link.source_link_ptr)
         { gzthrow("ModelTask: cannot find exported source link " << it->source_link << " in model"); }
         else if (exported_link.target_link != _world_frame.get() && !exported_link.target_link_ptr)
         { gzthrow("ModelTask: cannot find exported target link " << it->target_link << " in model"); }
         else if (it->port_name.empty())
         { gzthrow("ModelTask: no port name given in link export"); }
+        else if (it->port_period.isNull())
```
Comment (doudou) [inline tasks/ModelTask.cpp]: I think you should allow this, using the current "update at every simulation cycle" meaning.

File: tasks/ModelTask.cpp [hunk 3]
```diff
@@ -221,6 +224,17 @@ void ModelTask::updateLinks(base::Time const& time)
 {
     for(ExportedLinks::const_iterator it = exported_links.begin(); it != exported_links.end(); ++it)
     {
+        static std::map<std::string, base::Time> last_write_time;
```
Comment (doudou) [inline tasks/ModelTask.cpp]: Never. *EVER* use static variables until you are *WAY* more experienced than you are.

Static variables can never be reset, are not thread-safe, and are just plain not needed.

Just store the last_update information in the internal exported_links structure.

File: tasks/ModelTask.cpp (hunk 1 above)
Comment (doudou) [reply tasks/ModelTask.cpp:241]: Agreed. Good catch.

**Comment by doudou** [comment]
> Are we using them?

Not for the links, no. They are set by Syskit. I'll update the syskit side right now.

**Comment by doudou** [comment]
(Note, for testing, you can simply start gazebo standalone and then manually set the property and check the period)

**Comment by doudou** [comment]
Even better would be to write unit tests ... but I honestly don't think you have the time to start the test harness for this package.

File: tasks/ModelTask.cpp (hunk 2 above)
Comment (dbcesar) [reply tasks/ModelTask.cpp]: removed

File: tasks/ModelTask.cpp (hunk 3 above)
Comment (dbcesar) [reply tasks/ModelTask.cpp]: Good to know.

**Comment by dbcesar** [comment]
Almost done here, but first: 
Is dangerous to use `iterator` instead of `const_iterator` the `exported_link`?

>(Note, for testing, you can simply start gazebo standalone and then manually set the property and check the period)

How do I link the standalone gazebo with the rock plugin? I've never done that.

**Comment by doudou** [comment]
> How do I link the standalone gazebo with the rock plugin?

You use `rock-gazebo` instead of `gazebo` (or `rock-gzserver` instead of `gzserver`)

**Comment by doudou** [comment]
> Is dangerous to use iterator instead of const_iterator the exported_link?

It's absolutely not dangerous. `const` is really about semantics - showing that you don't intend to change a structure - more than safety.

**Comment by doudou** [comment]
You need to reset last_update on start. This is simulation, time can jump back between runs.

**Pull request merged**